#169257409 User Role Settings

RaymondMwaura opened the pull request:
#### What does this PR do?
This PR creates a role management system for the API

#### Description of Task to be completed?
- [x] setup a roles controller
- [x] setup roles routes
- [x] setup roles middleware
- [x] setup tests

#### How should this be manually tested?
- clone this branch to your workstation and navigate to the project folder using Terminal
- install dependencies with the command `npm install`
- update your database schema by running `npm run migrate:undo` and then `npm run migrate`
- run the seeder file to populate the database with a user with the 'super_administrator' role by running the command `npx sequelize-cli db:seed:all`. The user's details are as follows:
```
{ 
    email: 'super@administrator.com',
    password: '12345678',
}
``` 
Test 1
- Open Postman and make a login request using the 'super_administrator' credentials above to receive a token
- Append the token to the Authorization Header: Bearer Token
- Make a 'GET' request to the endpoint 'http://localhost:3000/api/v1/auth/user/'
- You will get a list of all registered users
- Signup a new user with the signup endpoint
- By default, this new user has a role 'requester'
- Attempt to retrieve the list of users again with this new (requester role) user, you will get an 'insufficient privileges' error

Test 2
- Using the 'super_administrator' user, change the role of the 'requester' user by making a 'PATCH' request on the endpoint 'http://localhost:3000/api/v1/auth/user/role' and make sure you append your token to the Authorization Header: Bearer Token
- Include the following details in the request body:
```
{
    email: theRequesterUserEmail,
    role: 'manager'
}
```
- You will receive a confirmation message that the role has been changed successfully
- Attempt the same action using an invalid token or a token from a 'non-super_administrator' user and you will get an error message

Test 3:
- Run integration tests using the command `npm test`

#### Any background context you want to provide?
We shall use this endpoint to restrict certain endpoints to users with specific roles

#### What are the relevant pivotal tracker stories?
[#169257409](https://www.pivotaltracker.com/story/show/169257409)

#### Screenshots (if appropriate)
<img width="1436" alt="allUsersSuccess" src="https://user-images.githubusercontent.com/17108099/69291056-aa06cc80-0c0a-11ea-8cd7-4fea96c0cfad.png">
<img width="1439" alt="changeRoleSuccess" src="https://user-images.githubusercontent.com/17108099/69291066-b5f28e80-0c0a-11ea-9baa-c50d90d328f9.png">

#### Questions:
N/A

b0nbon1 committed:
ft(one-way trip):A user should be able to make a way trip

- added the corresponding models files
- added the request one way trip controlle
- added the necessary validations
- added the tests
- added unit tests

[delivers #169257403]

timolinn commented on src/controllers/roles.js:
```diff
@@ -0,0 +1,38 @@
+    if (!user) {
+      return Responses.handleError(404, 'user not found', res);
+    }
+    await db.user.update(
```
Since we have a `UserService` here, I think we can move this functionality into the user service.

RaymondMwaura replied:
Changing it will break the code in the User's Controller. The methods 'forgotPassword' and 'findUser' use it in slightly different ways.

timolinn replied:
I meant create a new method in the UserService and make use it in the controller.

RaymondMwaura replied:
I've tried to create a new method in the service like this:
```
  static async findUserByEmailWithResponse(email) {
    const user = await db.user.findOne({
      where: { email }
    });
    if (!user) return false;
    return true;
  }
```

And then edited the controller like this:
```
  async changeRole(req, res) {
    if (UserServices.findUserByEmailWithResponse(req.body.email)) {
      await db.user.update(
        {
          role: req.body.role
        },
        { where: { email: req.body.email } }
      );
      // refetch user from database with updated role
      const updatedUser = await UserServices.findUserByEmail(req.body.email);
      // build returned object here:
      const userObject = {
        firstName: updatedUser.firstName,
        lastName: updatedUser.lastName,
        role: updatedUser.role,
      };
      return Responses.handleSuccess(200, 'user role updated', res, userObject);
    }
    return Responses.handleError(404, 'user not found', res);
  }
```

But that's not working. How can I go about this?

timolinn replied:
I was hoping you'd create a `setUserRole` method in the UserService and then move the role update part into it.

RaymondMwaura replied:
I've created the setUserRole. Check and see if it's what you had in mind.

RaymondMwaura replied:
I have now implemented this.

timolinn commented on src/controllers/roles.js:
```diff
@@ -0,0 +1,38 @@
+   * @returns {object} responses
+   */
+  async changeRole(req, res) {
+    const user = await UserServices.findUserByEmail(req.body.email);
```
Why `findUserByEmail` and not `findUserById `?

RaymondMwaura replied:
The story requirement is that you change a user's role based on their email.

timolinn replied:
That's ok

timolinn commented on src/routes/api/roles.js:
```diff
@@ -0,0 +1,106 @@
+ *         description: success
+ */
+router.get(
+  '/auth/user',
```
Make this plural ie `users` instead of `user`.

gildniy reviewed (changes requested):
Please resolve conflicts.

timolinn commented on src/controllers/roles.js:
```diff
@@ -0,0 +1,42 @@
+    } catch (error) {
+      return Responses.handleError(500, 'server error', res);
+    }
+  }
```
I think one `try - catch` block should be fine:

```javascript
try {
      const user = await UserServices.findUserByEmail(req.body.email);
      if (!user) {
        return Responses.handleError(404, 'user not found', res);
      }

      // Update User Role
      const updatedUser = await UserServices.setUserRole(user.email, req.body.role);

      const userObject = {
        firstName: updatedUser.firstName,
        lastName: updatedUser.lastName,
        role: updatedUser.role,
      };
      return Responses.handleSuccess(200, 'user role updated', res, userObject);
    } catch (error) {
      return Responses.handleError(error.statusCode || 500, error.message || 'server error', res);
    }
  }
```

RaymondMwaura replied:
I have removed all the try catch blocks on this method and added the catchErrors to my 'roles' routes with help from @b0nbon1

timolinn commented on src/services/User.service.js:
```diff
@@ -9,13 +9,30 @@ class UserServices {
+      );
+    } catch (error) {
+      return error;
+    }
```
Use the custom ErrorHandler:

```javascript
try {
    const user = await db.user.update(
             { role: req.body.role },
             { where: { email: req.body.email } 
    });
    return user; 
} catch (error) {
    throw new ErrorHandler(500, "an error occurred while setting role");
}
```
You should also return the updated `user` so as to prevent extra database call in the controller. It's generally a good idea to `throw` errors in your `services` and handle them in your controllers.

RaymondMwaura replied:
I have updated the method to return an object on update and included the error handler as well.

b0nbon1 commented:
@RaymondMwaura Great work, but I'm getting this error while trying to run the server. Please help
![image](https://user-images.githubusercontent.com/46062609/69406549-fc2f1700-0d0a-11ea-916c-1e4be649e0b8.png)

RaymondMwaura committed:
feat(setup-roles): implement role management system

- create models to handle role management
- create routes for creating roles and permissions
- create middleware to facilitate role management

[Delivers #169257409]

RaymondMwaura commented:
> @RaymondMwaura Great work, but I'm getting this error while trying to run the server. Please help
> ![image](https://user-images.githubusercontent.com/46062609/69406549-fc2f1700-0d0a-11ea-916c-1e4be649e0b8.png)

Try it out again. It is now rebased with the latest changes on develop. Make sure you have a super_administrator user first by running the seed file '20191112083122-demo-users.js'

nshaphan commented on seeders/20191112083122-demo-users.js:
```diff
@@ -4,9 +4,9 @@ const passwordHash = hash.generateSync('12345678');
+    email: 'super@administrator.com',
     password: passwordHash,
     createdAt: Sequelize.literal('CURRENT_TIMESTAMP'),
     updatedAt: Sequelize.literal('CURRENT_TIMESTAMP'),
```
You need to add a field `role: 'super_administrator'` because it is not possible now to change the user role.

RaymondMwaura replied:
Thanks for picking up on this Shaphan.

b0nbon1 replied:
And also set admin as verified

gildniy reviewed (dismissed):
Great work @RaymondMwaura it looks good to me, it may be merged!

RaymondMwaura commented:
> Great work @RaymondMwaura it looks good to me, it may be merged!

Kindly reapprove. I have added the role field to the seeders since it was missing as oberved by @nshaphan

nshaphan reviewed (dismissed):
Great works @RaymondMwaura , This works as expected.

gildniy reviewed (dismissed):
Looks good.

gildniy reviewed (dismissed):
Nice work @Raymond. It can be merged

b0nbon1 reviewed (dismissed):
tested and It is working correctly as needed. I think it is ready to be merged

timolinn reviewed (commented):
I noticed the `Booking` is missing after I pulled this branch?

gildniy commented on src/controllers/users.controller.js:
```diff
@@ -188,6 +188,24 @@ class UserController {
+        attributes: { exclude: ['password'] }
+      });
+      return Responses.handleSuccess(200, 'successfully retrieved all users', res, allUsers);
+    } catch (error) {
```
I think, the  `try - catch` is not need here, as we have this method is wrapped in the `catchErrors` in the route.

RaymondMwaura replied:
Thanks for that. I have removed it.

timolinn commented:
Works great! Good job @RaymondMwaura

b0nbon1 reviewed (approved):
well done, I have tested and Its working accordingly.

nshaphan reviewed (approved):
LGTM
